src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -160,16 +166,214 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
             }
         }
 
-        internal static void AddSegments(ArrayBuilder<SourceText> builder, SourceText text)
+        internal static void AddSegments(ArrayBuilder<SourceText> segments, SourceText text)
         {
             CompositeText composite = text as CompositeText;
             if (composite == null)
             {
-                builder.Add(text);
+                segments.Add(text);
             }
             else
             {
-                builder.AddRange(composite._texts);
+                segments.AddRange(composite._segments);
+            }
+        }
+
+        internal static SourceText ToSourceTextAndFree(ArrayBuilder<SourceText> segments, SourceText original, bool adjustSegments)
+        {
+            if (adjustSegments)
+            {
+                TrimInaccessibleText(segments);
+                ReduceSegmentCount(segments);
+            }
+
+            if (segments.Count == 0)
+            {
+                segments.Free();
+                return SourceText.From(string.Empty, original.Encoding, original.ChecksumAlgorithm);
+            }
+            else if (segments.Count == 1)
+            {
+                SourceText result = segments[0];
+                segments.Free();
+                return result;
+            }
+            else
+            {
+                return new CompositeText(segments.ToImmutableAndFree(), original.Encoding, original.ChecksumAlgorithm);
+            }
+        }
+
+        internal const int IDEAL_SEGMENT_COUNT = 10;
```

Comment:
Can you document (with XML docs) what these values mean, and why you chose them? "Made them up" is totally valid, but it's helpful if/when somebody wants to change it they know the backstory.